GSA/digitalgov.gov
# Rs hotfix/fix corrupt image uploads

## Summary
This PR refactors the file upload process to address issues with image uploads and file processing. It replaces the gulp-awspublish plugin with a custom TypeScript-based upload module using AWS SDK v3, while maintaining Gulp task integration for workflow consistency.


## Solution
This PR implements the following changes:

1. Removes the gulp-awspublish plugin and replaces it with a custom TypeScript module (`config/typescript/upload.ts`) using AWS SDK v3 for file uploads.
2. Maintains separate functions for uploading images and files, preserving the existing structure.
3. Implements a cleanup function to remove temporary working directories after successful uploads.
4. Updates the Gulpfile to integrate the new TypeScript-based upload module within the existing Gulp task structure.

This approach was chosen to:
1. Improve type safety and catch potential errors at compile-time by using TypeScript.
2. Modernize the upload process by utilizing the latest AWS SDK v3, removing the dependency on the outdated gulp-awspublish plugin.
3. Maintain the existing Gulp-based workflow for consistency and ease of integration.
4. Ensure proper cleanup after the upload process.

The change was implemented by creating a new TypeScript file for uploads, compiling it to JavaScript, and integrating it into the existing Gulp tasks. This allows us to have more control over the upload process while keeping the familiar Gulp task structure.

## How To Test
1. Install dependencies: `npm install`
2. Place test images in `content/uploads/_inbox/__add image or static files to this folder__`
3. Run the Gulp upload task: `npx gulp upload`
4. Verify that files are uploaded to the S3 bucket and local directories are cleaned up

## Dependency updates
### Dependency 
| Dev Dependency name               | Previous version | New version |
|-------------------------------|:----------------:|:-----------:|
| @aws-sdk/client-s3 |     --        |   ^3.614.0  |
| dotenv | -- | ^16.3.1 |
| mime-types | -- | ^2.1.35 |
### Dev Dependency
| Dev Dependency name               | Previous version | New version |
|-------------------------------|:----------------:|:-----------:|
| @typescript-eslint/eslint-plugin |     --        |   ^5.x.x    |
| @typescript-eslint/parser     |        --        |   ^5.x.x    |
| @types/mime-types | -- | ^2.1.4 | 
| typescript                    |        --        |   ^4.x.x    |
| gulp-awspublish               |     ^8.0.0       |   Removed   |

## Script updates
- Modified `lint:js` and `lint:js:fix` to include TypeScript files
- Modified `prettier:js` and `prettier:js:fix` to include TypeScript files
- Added `build` script to compile TypeScript
- Added `type-check` script to run TypeScript's type checker

## Notes
- TypeScript is now a local dependency; use `npm run build` to compile.
- ESLint configuration has been updated for TypeScript in `.eslintrc.yaml`.
- Run `npm run build`, `npm run lint:js`, and `npm run type-check` before submitting PR to ensure code quality.
